### PR TITLE
Zero and identity matrix

### PR DESCRIPTION
**Pull request by wbhart**
* Remove eye
* Add identity_matrix(R, m, n) where m != n, setting ones down the diagonal
* Add zero_matrix(M), identity_matrix(M) where M is a matrix
* Add zero_matrix(M, n), identity_matrix(M, n) where M is a matrix

The final one is not currently documented. I can document it if you think it is desirable (I was not intending to implement it, but found it necessary after removing eye, as hnf seems to need it).

Also, do you want me to add a 2 integer version of that last one?

### CODE REVIEW

**Comment by wbhart** [comment]
Closes #435

**Comment by rfourquet** [comment]
Isn't `zero_matrix(matrix)` doing what `zero(matrix)` is already doing?

**Comment by wbhart** [comment]
It probably is. I'll take a look and remove one of the implementations.

Of course zero is essentially defined only for ring elements in AbstractAlgebra, so we'd still want zero_matrix for matrix spaces (and it doesn't hurt to have it anyway).

**Comment by wbhart** [comment]
It looks like the one integer version of zero is missing for matrix algebras though, so I can't quite implement all these in terms of existing functions.

**Comment by wbhart** [comment]
Actually, what I'll do is leave out both implementations of zero_matrix that hnf relies on and use zero instead. That will simplify my PR.

**Comment by rfourquet** [comment]
> It looks like the one integer version of zero is missing for matrix algebras though

`zero(mat, n)` is defined if `mat isa MatAlgElem`, but indeed not if `m` is a `MatElem`.

**Comment by wbhart** [comment]
I'm so confused now. It seems like we are using zero and one for things that aren't rings now, like matrix spaces. But if you want to create a zero matrix or matrix with ones down the diagonal, zero and one won't work because they don't take the base ring. Therefore we have zero_matrix and identity_matrix (even though it's not an identity). But the identity_matrix I just implemented violates this.

Anyhow, I removed the zero_matrix(M, ...) functions, since they just duplicate zero functionality.

**Comment by wbhart** [comment]
Let's pose this as a series of questions for @thofma and @fieker 

1) Do we really want zero and one to apply to things that are not rings? E.g. matrix spaces.

2) Do we really want identity_matrix to create matrices that are not the identity, or should we pick a better name? (Note, we can't use one, because it takes the ring, not the matrix space as an argument.)

3) Assuming we want identity_matrix(M) where M is a matrix (this replaces eye, which I removed), do we also want zero_matrix(M). (Note that these would be duplicating what zero and one currently do.)

4) Assuming we want identity_matrix(M) and presumably identity_matrix(M, n), (which seems to be needed by HNF), do we also want identity_matrix(M, r, c) where M is a matrix? Similar question for zero_matrix.

5) Is the Easter Bunny real (we have to have at least one question on which we can all agree on the answer).

My votes are as follows:

1) No.
2) No, we should pick a better name.
3) Don't care, so long as there isn't one without the other.
4) Yes.
5) Yes.

**Comment by fieker** [comment]
On Mon, Oct 07, 2019 at 10:43:18AM -0700, wbhart wrote:
> Let's pose this as a series of questions for @thofma and @fieker 
> 
> 1) Do we really want zero and one to apply to things that are not rings? E.g. matrix spaces.
zero to a matrix (space) makes sense to me, I cannot understand the idea
between identity_matirx(ZZ, 2, 3), but if you support that, then one is
also fine.
> 
> 2) Do we really want identity_matrix to create matrices that are not the identity, or should we pick a better name? (Note, we can't use one, because it takes the ring, not the matrix space as an argument.)
I still don't see non-square identity matrices, but there might be a
point. I recall, the key support is that julia, thus the numerical
crowd, seem to live by this
> 
> 3) Assuming we want identity_matrix(M) where M is a matrix (this replaces eye, which I removed), do we also want zero_matrix(M). (Note that these would be duplicating what zero and one currently do.)
its a question of consistency, in both directions.
one(R, 3) to me, is non-sensical, while identity_matrix(R, 3) is
expected
BUT
one(MatrixSpace()) is also expected, one(MatrixAlgebra) is even required

In fact, one(R, 1) could also be an all-1-array of length 3...

Identical comments apply to zero
> 
> 4) Assuming we want identity_matrix(M) and presumably identity_matrix(M, n), (which seems to be needed by HNF), do we also want identity_matrix(M, r, c) where M is a matrix? Similar question for zero_matrix.
what is M here? I still don't see non-square identity matrices, but if
you have an example (a good one) be my guest. (I also don't feel
strongly against it)

> 
> 5) Is the Easter Bunny real (we have to have at least one question on which we can all agree on the answer).
> 
> My votes are as follows:
> 
> 1) No.
> 2) No, we should pick a better name.
> 3) Don't care, so long as there isn't one without the other.
> 4) Yes.
> 5) Yes.
> 
Look, duplication is not always bad. Functions need to be discoverable
by normal (?) users that do not know us in person. This means that we
wil lpossible have to have a couple of functions under multiple names.
It sucks since this is non-trivial to maintain, but to me this is
unavoidable.
(Non trivial:
  thinking of maximal_order = ring_of_integers = integers in number
  theory,
  this would not make any sense where maximal_order would return the
  maximal order of say, any element in a given group.
  There are better examples than this, but I currently cannot recall
  them. I ran into this at magma.)

What is the downside of having one, zero as well as identity_matrix and
zero_matrix? My main concern would be that one(R, 3, 4) I'd expect an
all-1-matrix, not an identity one.

**Comment by rfourquet** [comment]
> 1. Do we really want zero and one to apply to things that are not rings? E.g. matrix spaces.

`zero` makes sense for a group, `one` can make sense to a `MatElem` matrix which is square. 

> 2. Do we really want identity_matrix to create matrices that are not the identity, or should we pick a better name? 

In Julia, one can use `LinearAlgebra.I`:
```julia
julia> Matrix(I, 2, 3)
2×3 Array{Bool,2}:
 1  0  0
 0  1  0
```
By analogy, we could support `matrix(ring, 2, 3, I)`. Currently, we already have an indirect way for square matrices:
```julia
julia> matrix(ZZ, collect(BigInt, I(2)))
[1  0]
[0  1]
```

**Comment by thofma** [comment]
1) Don't mind.
2) I like the functionality and can live that name, but I don't have strong feelings.
3) I don't want these functions, but I don't have any strong feelings.
4) As 3).

**Comment by wbhart** [comment]
Nobody seems to have very strong feelings, and @thofma can live with the PR as it is, which is less work than me.

The original intention was to support identity_matrix(R, 3, 3) and to get rid of eye, finally. The PR achieves that.

I think we are all a bit nervous about also supporting identity_matrix(R, 3, 2), but I also don't see an overwhelming reason to ban it. So I'll leave it as is, and if this causes someone a real problem, we'll raise an exception.

As for one(M) returning all ones, I would prefer ones(M) to do that, if we ever have such a function. I would expect one(M) to return an "identity matrix", for sure. So let's stick with that.

As for supporting I, I am not in favour. It's just namespace pollution, and particularly bad pollution for an algebraist. So that for sure stays out.

So in short, I think this PR is ready to go. I can live with it if everyone else can.
